apps/prairielearn/elements/pl-file-upload/pl-file-upload.js

```diff
@@ -83,26 +105,33 @@
         accept: (file, done) => {
           // fuzzy case match
           const fileNameLowerCase = file.name.toLowerCase();
-          if (this.acceptedFilesLowerCase.includes(fileNameLowerCase)) {
+          if (this.findAcceptedFileName(fileNameLowerCase)) {
             return done();
           }
           return done('invalid file');
         },
         addedfile: (file) => {
+          const existingFileSize = this.files.reduce((prev, next) => prev + next.size, 0);
+          if (existingFileSize + file.size > this.maxFileSizeMB * 1024 * 1024) {
+            this.addWarningMessage(
+              `Combined file size of new file and existing files (<strong>${
+                Math.round((existingFileSize + file.size) / 1024 / 10.24) / 100
+              } MB</strong>) and is greater than maximum file size of ${this.maxFileSizeMB} MB.`,
+            );
+            return;
+          }
           // fuzzy case match
           const fileNameLowerCase = file.name.toLowerCase();
-          if (!this.acceptedFilesLowerCase.includes(fileNameLowerCase)) {
+          const acceptedFileName = this.findAcceptedFileName(fileNameLowerCase);
+
+          if (acceptedFileName === null) {
             this.addWarningMessage(
-              '<strong>' +
-                file.name +
-                '</strong>' +
-                ' did not match any accepted file for this question.',
+              `<strong>${file.name}</strong> did not match any accepted file for this question.`,
```

Comment:
Upon further thought, since this is only done on the student's client side, and this message doesn't persist to instructor side, I'm guessing no, there's no issue here specifically. That said, it's probably a good idea to double check other places where file names are used without escaping.

Reply:
I don't think file names are sanitized at any point even in the existing code, but I guess this wouldn't be an issue in the old implementation because names are much more restrictive.

Reply:
I finally had some time to look into this, and I think the best way to solve this is to simply reject file names that contain < or >.

Here's the alternatives I considered and why I think they are problematic:
1. Sanitizing file names everywhere they are printed is tedious and also quite dangerous since this might happen in a lot of places (not just in this element), and it'll be easy to miss some special case (e.g., I'm not 100% sure if there could be file names included in certain error messages).
2. Sanitizing file names as they are added creates some awkwardness since the files might technically no longer match the pattern after sanitizing (e.g., ">.py" matches the pattern "?.py", but "&gt;.py" does not). This could technically be worked around, but might also lead to some display issues down the road where "&gt;.py" is displayed as such and not as HTML, which might confuse students/instructors.
3. I also considered stricter banning of characters (single and double quotes, as well as the & character are typically problematic), but those are much more likely to be used in actual file names without any malicious intent, and as far as I can see, we never run into a case where these characters would cause trouble during printing (URLs are already escaped in the existing code, and we use IDs rather than file names in HTML attributes).

Note that this could lead to (I assume very rare) backward compatibility issues if an instructor actually used a file name with < or > in the `file-names` attribute. However, this would have likely caused some display issues before this PR (at the very least, file names might not have been displayed correctly), so I personally don't feel too bad about those cases. If others think that this is a problem, I could allow < or > in mandatory file names only.

Reply:
(Ignore the previous comment; the idea of just blocking < and > is not sufficient as it still allows display shenanigans like "\&nsbp;.py", and I also didn't realize that file-preview already has working escaping).

I checked and as far as I can see, this is the only place where escaping actually needs to be added explicitly is the JS code for file-upload (so this is only a client-side issue, as @jonatanschroeder initially suspected). In the file-preview, Mustache is already escaping the file names. 

I added the escaping now and everything seems to be displayed correctly now across both the file-upload and file-preview element.